### PR TITLE
Improve open_in_browser to handle missing head tags

### LINKED ISSUE #6550: Improve open_in_browser function's handling of base tags in response.py
The `open_in_browser` function in `response.py` uses a workaround to insert a `<base>` tag into the HTML response for local file viewing. As mentioned in the function's docstring, is "dirty" and could fail in situations where `<head>` tags are missing or improperly formed which can lead to unintended effects when rendering HTML.

Steps to Reproduce:
1. Use `open_in_browser` with an HTML response that does not contain a proper `<head>` tag.
2. Observe to see if the function does not handle this smoothly. 

Suggested Fix:
Edit the function to handle missing or properly formed `<head>` tags so that a correct `<base>` tag can be inserted.

### PR DESCRIPTION
**Pull request by vassilismylo**
# Fix open_in_browser to handle missing head tags

This PR improves the `open_in_browser` function to properly handle HTML responses that don't have a `<head>` tag. Previously, the function would fail to insert the `<base>` tag in such cases, causing relative links to break when viewing the page in a browser.

## Changes

- Modified the `open_in_browser` function to check for several HTML structure patterns
- Added a fallback mechanism that inserts both `<head>` and `<base>` tags when needed
- Implemented a cascading approach: first try `<head>`, then `<body>`, then `<html>`, and finally prepend to document

## Benefits

- More robust handling of various HTML structures
- Ensures relative URLs work properly in all viewed responses
- No new dependencies added

## Testing

Added new test cases that cover:
- HTML without head tags
- HTML with only body tags
- HTML with only html tags
- Minimal HTML with no structural tags

Fixes #6550

### CODE REVIEW

**Comment by vassilismylo** [comment]
fixing the failing check now

File: scrapy/utils/response.py [hunk 1]
```diff
@@ -93,14 +93,36 @@ def parse_details(self, response):
...
+                    html_match = re.search(rb"<html(?:[^<>]*?>)", body)
+                    if html_match:
+                        repl = rf"\0<head>{base_tag}</head>"
+                        body = re.sub(
+                            rb"<html(?:[^<>]*?>)", to_bytes(repl), body, count=1
+                        )
+                    else:
+                        # As a last resort, prepend to document
+                        body = to_bytes(f"<head>{base_tag}</head>") + body
+            ext = ".html"
```
Comment (wRAR) [inline scrapy/utils/response.py:125]: This line is duplicated.

File: scrapy/utils/response.py [hunk 2]
```diff
@@ -93,14 +93,36 @@ def parse_details(self, response):
...
         if b"<base" not in body:
             _remove_html_comments(body)
-            repl = rf'\0<base href="{response.url}">'
-            body = re.sub(rb"<head(?:[^<>]*?>)", to_bytes(repl), body, count=1)
-        ext = ".html"
+            base_tag = f'<base href="{response.url}">'
+
+            # Try to insert after <head> tag first (existing behavior)
+            head_match = re.search(rb"<head(?:[^<>]*?>)", body)
+            if head_match:
```
Comment (wRAR) [inline scrapy/utils/response.py:105]: I would rewrite these checks to use `:=`

File: tests/test_utils_response.py [hunk 3]
```diff
@@ -204,6 +205,64 @@ def test_open_in_browser_redos_head(self):
         assert end_time - start_time < MAX_CPU_TIME
 
 
+class TestOpenInBrowserWithoutHead(unittest.TestCase):
```
Comment (wRAR) [inline tests/test_utils_response.py:208]: I don't think this needs to inherit from `unittest.TestCase`

Comment (vassilismylo) [reply tests/test_utils_response.py:208]: thank for replying, will fix these and create a new pull request

**Comment by vassilismylo** [comment]
@wRAR  Thanks for the review! I've made the following improvements based on your feedback:

**Code style improvements:**
- Replaced nested if/else with clean elif chain for better readability
- Simplified regex patterns (removed unnecessary non-capturing groups)
- Removed unused walrus operator variables that were triggering linting errors - this is the commit message

- Unfortunatelly i did not write a correct commit message:
**Test improvements:**
- Test class no longer inherits from unittest.TestCase
- Using pathlib.Path instead for file operations
- Added comprehensive test coverage for all edge cases

File: scrapy/utils/response.py [hunk 4]
```diff
@@ -93,13 +93,27 @@ def parse_details(self, response):
...
-            repl = rf'\0<base href="{response.url}">'
-            body = re.sub(rb"<head(?:[^<>]*?>)", to_bytes(repl), body, count=1)
+            body = _remove_html_comments(body)
+            base_tag = f'<base href="{response.url}">'
+
+            # Try to insert after <head> tag first (existing behavior)
+            if re.search(rb"<head[^<>]*?>", body):
+                repl = rf"\0{base_tag}"
+                body = re.sub(rb"<head[^<>]*?>", to_bytes(repl), body, count=1)
+            # No head tag found, try to insert at beginning of body
```
Comment (wRAR) [inline scrapy/utils/response.py:106]: I think this inserts `head` inside `body` but it should be inserted before it, not inside.

File: tests/test_utils_response.py [hunk 5]
```diff
@@ -204,6 +204,91 @@ def test_open_in_browser_redos_head(self):
...
+        """Test with HTML that doesn't have a head tag"""
+        body = b"""<html>
+            <body><p>Hello world.</p></body>
+        </html>"""
+        response = HtmlResponse(self.url, body=body)
+
+        # Mock _openfunc to capture the written content
+        written_content = []
+
+        def mock_open(url):
```
Comment (wRAR) [inline tests/test_utils_response.py:225]: This function is repeated, can it be extracted?

File: tests/test_utils_response.py [hunk 6]
```diff
@@ -204,6 +204,91 @@ def test_open_in_browser_redos_head(self):
         assert end_time - start_time < MAX_CPU_TIME
 
 
+class TestOpenInBrowserWithoutHead:
+    def setup_method(self):
+        self.url = "http://example.com"
```
Comment (wRAR) [inline tests/test_utils_response.py:209]: This can be a class attr

**Comment by AdrianAtZyte** [comment]
Closing due to a lack of feedback.
